utility_analysis_new/tests/parameter_tuning_test.py

```diff
@@ -279,6 +290,66 @@ def test_quantile_contributions(self, bins, q, expected_quantiles):
         output = histogram.quantiles(q)
         self.assertListEqual(expected_quantiles, output)
 
+    @parameterized.parameters(
+        (True, True, [1, 1, 2, 2, 6, 6], [3, 6, 3, 6, 3, 6]),
+        (False, True, None, [3, 6]), (True, False, [1, 2, 6], None))
+    def test_find_candidate_parameters(
+        self,
+        tune_max_partitions_contributed: bool,
+        tune_max_contributions_per_partition: bool,
+        expected_max_partitions_contributed: List,
+        expected_max_contributions_per_partition: List,
+    ):
+        mock_l0_histogram = parameter_tuning.Histogram(None, None)
+        mock_l0_histogram.quantiles = mock.Mock(return_value=[1, 1, 2])
+        setattr(mock_l0_histogram.__class__, 'max_value', 6)
+        mock_linf_histogram = parameter_tuning.Histogram(None, None)
+        mock_linf_histogram.quantiles = mock.Mock(return_value=[3, 6, 6])
+
+        mock_histograms = parameter_tuning.ContributionHistograms(
+            mock_l0_histogram, mock_linf_histogram)
+        parameters_to_tune = parameter_tuning.ParametersToTune(
+            max_partitions_contributed=tune_max_partitions_contributed,
+            max_contributions_per_partition=tune_max_contributions_per_partition
+        )
+
+        candidates = parameter_tuning._find_candidate_parameters(
+            mock_histograms, parameters_to_tune)
+        self.assertEqual(expected_max_partitions_contributed,
+                         candidates.max_partitions_contributed)
+        self.assertEqual(expected_max_contributions_per_partition,
+                         candidates.max_contributions_per_partition)
+
+    def test_tune(self):
+        input = [(i % 10, f"pk{i/10}") for i in range(10)]
+        public_partitions = [f"pk{i}" for i in range(10)]
+        data_extractors = pipeline_dp.DataExtractors(
+            privacy_id_extractor=lambda x: x[0],
+            partition_extractor=lambda x: x[1],
+            value_extractor=lambda x: None)
+
+        contribution_histograms = list(
```

Comment:
nit: just for my understanding, the user won't have to call  parameter_tuning.compute_contribution_histograms() explicitly, this will be done for them by some script, is this right?

Reply:
calling of `compute_contribution_histograms` should be done by some caller code outside of PipelineDP. It's because it's framework specific :(